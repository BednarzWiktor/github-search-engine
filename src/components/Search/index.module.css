.container, .option, .filters {
    width: 100%;
}

.option, .filters {
    box-sizing: border-box;
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.filters {
    padding: 0 1rem;
    margin: 0 0 1rem;
}