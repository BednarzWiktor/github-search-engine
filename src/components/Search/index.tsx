import React, { useState, useEffect, SyntheticEvent } from 'react';

import { debounce } from 'lodash';

import { TextField } from '@material-ui/core';
import { Autocomplete } from '@material-ui/lab';
import OptionIconGenerator from '../OptionIconGenerator';

import { SearchProps as Props, SearchResultItem } from '../../types/propTypes';

import styles from './index.module.css';

const Search = ({ combinedResults, getSearchResults }: Props) => {
    const [ value, setValue ] = useState<SearchResultItem | null>(null);
    const [ inputValue, setInputValue ] = useState<string>('');

    const handleOnChangeValue = (event: SyntheticEvent<EventTarget>, newValue: any) => {
        setValue(newValue);
    };

    const handleSearchWithDebounce = debounce((searchValue) => {
        if (searchValue && searchValue.length >= 3) {
            getSearchResults(searchValue);
        }
    }, 300);

    const handleOnChangeInputValue = (event: SyntheticEvent<EventTarget>, newInputValue: string) => {
        setInputValue(newInputValue);
        handleSearchWithDebounce(newInputValue)
    };

    useEffect(() => {
        if (value) {
            window.open(value.url, "_blank");
        }
    }, [ value ]);

    return (
        <div className={styles.container}>
            <span className={styles.input}>
                <Autocomplete
                    value={value}
                    onChange={handleOnChangeValue}
                    inputValue={inputValue}
                    onInputChange={handleOnChangeInputValue}
                    options={combinedResults}
                    getOptionLabel={item => item.name}
                    clearOnBlur={false}
                    renderInput={
                        props =>
                            <TextField
                                {...props}
                                label="Search GitHub"
                                variant="outlined"
                            />
                    }
                    renderOption={
                        option =>
                            <span className={styles.option}>
                                <span>{option.name}</span>
                                <OptionIconGenerator origin={option.origin}/>
                            </span>
                    }
                    onClose={(event, reason) => console.log(reason)}
                />
            </span>
        </div>
    );
};

export default Search;