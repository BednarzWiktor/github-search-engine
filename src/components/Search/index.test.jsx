import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import toJson from 'enzyme-to-json';

import Search from './';

configure({adapter: new Adapter()});
describe('Search', () => {
    const basicProps = {
        combinedResults: [
            {
                name: 'user1',
                url: 'http://user1.com',
                origin: 'user'
            },
            {
                name: 'repo1',
                url: 'http://repo1.com',
                origin: 'user'
            },
            {
                name: 'user2',
                url: 'http://user2.com',
                origin: 'user'
            },
            {
                name: 'repo2',
                url: 'http://repo2.com',
                origin: 'user'
            },
        ],
        getSearchResults: jest.fn(),
        clearSearchResults: jest.fn(),
    };
    const isLoadingProps = { ...basicProps, isLoading: true };
    const emptyProps = { ...basicProps, combinedResults: [] };
    const emptyIsLoadingProps = { ...emptyProps, isLoading: true };

    const basicWrapper = shallow(<Search { ...basicProps }/>);
    const isLoadingWrapper = shallow(<Search { ...isLoadingProps }/>);
    const emptyWrapper = shallow(<Search { ...emptyProps }/>);
    const emptyIsLoadingWrapper = shallow(<Search { ...emptyIsLoadingProps }/>);

    it('renders without crashing', () => {
        expect(toJson(basicWrapper)).toMatchSnapshot();
        expect(toJson(isLoadingWrapper)).toMatchSnapshot();
        expect(toJson(emptyWrapper)).toMatchSnapshot();
        expect(toJson(emptyIsLoadingWrapper)).toMatchSnapshot();
    });
});