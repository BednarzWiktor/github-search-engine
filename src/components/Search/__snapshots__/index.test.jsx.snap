// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders without crashing 1`] = `
<div
  className="container"
>
  <aside
    className="filters"
  >
    <WithStyles(ForwardRef(Typography))
      component="span"
      variant="body1"
    >
      Search within:
    </WithStyles(ForwardRef(Typography))>
    <span
      className="filtersControls"
    >
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Users"
            onChange={[Function]}
          />
        }
        label="Users"
      />
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Repos"
            onChange={[Function]}
          />
        }
        label="Repos"
      />
    </span>
  </aside>
  <main
    className="input"
  >
    <WithStyles(ForwardRef(Autocomplete))
      clearOnBlur={false}
      disabled={false}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      inputValue=""
      loading={false}
      loadingText="Fetching data..."
      onChange={[Function]}
      onInputChange={[Function]}
      options={
        Array [
          Object {
            "name": "user1",
            "origin": "user",
            "url": "http://user1.com",
          },
          Object {
            "name": "repo1",
            "origin": "user",
            "url": "http://repo1.com",
          },
          Object {
            "name": "user2",
            "origin": "user",
            "url": "http://user2.com",
          },
          Object {
            "name": "repo2",
            "origin": "user",
            "url": "http://repo2.com",
          },
        ]
      }
      renderInput={[Function]}
      renderOption={[Function]}
      value={null}
    />
  </main>
  <WithStyles(ForwardRef(Snackbar))
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Alert))
      elevation={6}
      onClose={[Function]}
      severity="error"
      variant="filled"
    />
  </WithStyles(ForwardRef(Snackbar))>
</div>
`;

exports[`Search renders without crashing 2`] = `
<div
  className="container"
>
  <aside
    className="filters"
  >
    <WithStyles(ForwardRef(Typography))
      component="span"
      variant="body1"
    >
      Search within:
    </WithStyles(ForwardRef(Typography))>
    <span
      className="filtersControls"
    >
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Users"
            onChange={[Function]}
          />
        }
        label="Users"
      />
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Repos"
            onChange={[Function]}
          />
        }
        label="Repos"
      />
    </span>
  </aside>
  <main
    className="input"
  >
    <WithStyles(ForwardRef(Autocomplete))
      clearOnBlur={false}
      disabled={false}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      inputValue=""
      loading={true}
      loadingText="Fetching data..."
      onChange={[Function]}
      onInputChange={[Function]}
      options={
        Array [
          Object {
            "name": "user1",
            "origin": "user",
            "url": "http://user1.com",
          },
          Object {
            "name": "repo1",
            "origin": "user",
            "url": "http://repo1.com",
          },
          Object {
            "name": "user2",
            "origin": "user",
            "url": "http://user2.com",
          },
          Object {
            "name": "repo2",
            "origin": "user",
            "url": "http://repo2.com",
          },
        ]
      }
      renderInput={[Function]}
      renderOption={[Function]}
      value={null}
    />
  </main>
  <WithStyles(ForwardRef(Snackbar))
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Alert))
      elevation={6}
      onClose={[Function]}
      severity="error"
      variant="filled"
    />
  </WithStyles(ForwardRef(Snackbar))>
</div>
`;

exports[`Search renders without crashing 3`] = `
<div
  className="container"
>
  <aside
    className="filters"
  >
    <WithStyles(ForwardRef(Typography))
      component="span"
      variant="body1"
    >
      Search within:
    </WithStyles(ForwardRef(Typography))>
    <span
      className="filtersControls"
    >
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Users"
            onChange={[Function]}
          />
        }
        label="Users"
      />
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Repos"
            onChange={[Function]}
          />
        }
        label="Repos"
      />
    </span>
  </aside>
  <main
    className="input"
  >
    <WithStyles(ForwardRef(Autocomplete))
      clearOnBlur={false}
      disabled={false}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      inputValue=""
      loading={false}
      loadingText="Fetching data..."
      onChange={[Function]}
      onInputChange={[Function]}
      options={Array []}
      renderInput={[Function]}
      renderOption={[Function]}
      value={null}
    />
  </main>
  <WithStyles(ForwardRef(Snackbar))
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Alert))
      elevation={6}
      onClose={[Function]}
      severity="error"
      variant="filled"
    />
  </WithStyles(ForwardRef(Snackbar))>
</div>
`;

exports[`Search renders without crashing 4`] = `
<div
  className="container"
>
  <aside
    className="filters"
  >
    <WithStyles(ForwardRef(Typography))
      component="span"
      variant="body1"
    >
      Search within:
    </WithStyles(ForwardRef(Typography))>
    <span
      className="filtersControls"
    >
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Users"
            onChange={[Function]}
          />
        }
        label="Users"
      />
      <WithStyles(ForwardRef(FormControlLabel))
        control={
          <ForwardRef(WithStyles)
            checked={true}
            color="primary"
            name="Repos"
            onChange={[Function]}
          />
        }
        label="Repos"
      />
    </span>
  </aside>
  <main
    className="input"
  >
    <WithStyles(ForwardRef(Autocomplete))
      clearOnBlur={false}
      disabled={false}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      inputValue=""
      loading={true}
      loadingText="Fetching data..."
      onChange={[Function]}
      onInputChange={[Function]}
      options={Array []}
      renderInput={[Function]}
      renderOption={[Function]}
      value={null}
    />
  </main>
  <WithStyles(ForwardRef(Snackbar))
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Alert))
      elevation={6}
      onClose={[Function]}
      severity="error"
      variant="filled"
    />
  </WithStyles(ForwardRef(Snackbar))>
</div>
`;
